lib/CL/pocl_cache.c

```diff
@@ -0,0 +1,470 @@
+    unsigned i;
+    cl_device_id device = program->devices[device_i];
+
+    pocl_SHA1_Init(&hash_ctx);
+
+    if (program->source) {
+        assert(preprocessed_source);
+        assert(source_len > 0);
+        pocl_SHA1_Update(&hash_ctx, (uint8_t*)preprocessed_source,
+                         source_len);
+    } else     { /* Program was created with clCreateProgramWithBinary() */
+        assert(program->binaries[device_i]);
+        pocl_SHA1_Update(&hash_ctx,
+                         (uint8_t*) program->binaries[device_i],
+                         program->binary_sizes[device_i]);
+    }
+
+    if (program->compiler_options)
+        pocl_SHA1_Update(&hash_ctx, (uint8_t*) program->compiler_options,
+                         strlen(program->compiler_options));
+
+    /* The kernel compiler work-group function method affects the
+       produced binary heavily. */
+    const char *wg_method=
+        pocl_get_string_option("POCL_WORK_GROUP_METHOD", "");
+
+    pocl_SHA1_Update(&hash_ctx, (uint8_t*) wg_method, strlen(wg_method));
+    pocl_SHA1_Update(&hash_ctx, (uint8_t*) PACKAGE_VERSION,
+                     strlen(PACKAGE_VERSION));
+    pocl_SHA1_Update(&hash_ctx, (uint8_t*) LLVM_VERSION,
+                     strlen(LLVM_VERSION));
+    pocl_SHA1_Update(&hash_ctx, (uint8_t*) POCL_BUILD_TIMESTAMP,
+                     strlen(POCL_BUILD_TIMESTAMP));
+    pocl_SHA1_Update(&hash_ctx, (const uint8_t *)POCL_KERNELLIB_SHA1,
+                     strlen(POCL_KERNELLIB_SHA1));
+    /*devices may include their own information to hash */
+    if (device->ops->build_hash)
+        device->ops->build_hash(device->data, &hash_ctx);
+
+
+    uint8_t digest[SHA1_DIGEST_SIZE];
+    pocl_SHA1_Final(&hash_ctx, digest);
+
+    unsigned char* hashstr = program->build_hash[device_i];
+    for (i=0; i < SHA1_DIGEST_SIZE; i++)
+        {
+            *hashstr++ = (digest[i] & 0x0F) + 65;
+            *hashstr++ = ((digest[i] & 0xF0) >> 4) + 65;
+        }
+    *hashstr = 0;
+
+    program->build_hash[device_i][2] = '/';
+
+}
+
+#ifdef ANDROID
+static
+char* pocl_get_process_name()
+{
+    char tmpStr[64], cmdline[512], *processName=NULL;
+    FILE *statusFile;
+    size_t len, i, begin;
+
+    snprintf(tmpStr, 64, "/proc/%d/cmdline", getpid());
+    statusFile=fopen(tmpStr, "r");
+    if (statusFile == NULL)
+        return NULL;
+
+    if (fgets(cmdline, 511, statusFile) != NULL) {
+        len=strlen(cmdline);
+        begin=0;
+        for (i=len-1; i >= 0; i--) { /* Extract program-name after last '/' */
+            if (cmdline[i] == '/') {
+                begin=i + 1;
+                break;
+            }
+        }
+        processName=strdup(cmdline + begin);
+    }
+
+    fclose(statusFile);
+    return processName;
+}
+#endif
+
+
+/******************************************************************************/
+
+static void pocl_cache_init_topdir() {
+
+    char *tmp_path=NULL;
+
+    tmp_path = getenv("POCL_CACHE_DIR");
```

Comment:
Use pocl_runtime_config.c for envs. We might add a configuration file or similar method to change these parameters later on.